megamek/src/megamek/common/ConstructionFactorWarning.java

```diff
@@ -0,0 +1,201 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with MegaMek. If not, see <http://www.gnu.org/licenses/>.
+ */
+package megamek.common;
+
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+
+import org.apache.logging.log4j.LogManager;
+
+import megamek.client.ui.swing.DeploymentDisplay;
+import megamek.client.ui.swing.GUIPreferences;
+import megamek.client.ui.swing.MovementDisplay;
+import megamek.common.enums.GamePhase;
+
+/**
+ * Construction Factor Warning Logic.  Handles events, help
+ * methods and logic related to CF Warning in a way that
+ * can be unit tested and encapsulated from BoardView and
+ * ClientGUI and other actors.
+ */
+public class ConstructionFactorWarning {
+    /*
+     * Handler for ClientGUI actionPerformed event. Encapsulates
+     * as much Construction Factory Warning logic possible.
+     */
+    public static void handleActionPerformed() {
+        toggleCFWarning();
+    }
+
+    /*
+     * Return true if the passed in phase is a phase that should allow
+     * Construction Factor Warnings such as Deploy and Movement.
+     */
+    public static boolean isCFWarningPhase(GamePhase gp) {
+        return (gp == GamePhase.DEPLOYMENT || gp == GamePhase.MOVEMENT);
+    }
+
+    public static boolean shouldShow(GamePhase gp) {
+        return shouldShow(gp, true);
+    }
+
+    /*
+     * Returns true if the show construction factor warning preference
+     * is enabled and in a phase that should show warnings.
+     */
+    public static boolean shouldShow(GamePhase gp, boolean isEnabled) {
+        return (isEnabled && isCFWarningPhase(gp));
+    }
+
+    private static boolean toggleCFWarning() {
+        //Toggle the GUI Preference setting for CF Warning setting.
+        GUIPreferences GUIP = GUIPreferences.getInstance();
+        GUIP.setShowCFWarnings(!GUIP.getShowCFWarnings());
+        return (GUIP.getShowCFWarnings());
+    }
+
+    /**
+     * For the provided entity, find all hexes within movement range with
+     * buildings that would collapse if entered or landed upon.  This is
+     * used by the {@link MovementDisplay} class.
+     *
+     * @param g {@link Game} provided by the phase display class
+     * @param e {@link Entity} currently selected in the movement phase.
+     * @param b {@link Board} board object with building data.
+     *
+     * @return returns a list of {@link Coords} that where warning flags
+     * 		should be placed.
+     */
+    public static List<Coords> findCFWarningsMovement(Game g, Entity e, Board b) {
+        List<Coords> warnList = new ArrayList<Coords>();
+
+        try {
+            //Only calculate CF Warnings for entity types in the whitelist.
+            if (!entityTypeIsInWhiteList(e)) {
+                return warnList;
+            }
+
+            Coords pos = e.getPosition();
+            int range = (e.getJumpMP() > e.getRunMP()) ? e.getJumpMP() : e.getRunMP();
```

Comment:
I find this more readable, but I consider that preference.
```suggestion
            int range = Math.max(e.getJumpMP(), e.getRunMP());
```

Reply:
Of course! If I wasn't drinking "eggnog", I would have thought of this.

Reply:
Updated.